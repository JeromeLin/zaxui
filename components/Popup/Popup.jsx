import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Events from '../utils/events';
import Mask from '../Mask';

class Popup extends PureComponent {

  constructor(props) {
    super(props);

    this.state = {
      isShow: this.props.visible || false,
      timer: null,
    };

    this.animationEnd = this.animationEnd.bind(this);
  }

  componentDidMount() {
    Events.on(this.popup, 'webkitTransitionEnd', this.animationEnd);
    Events.on(this.popup, 'webkitTransitionEnd', this.animationEnd);
  }

  componentWillReceiveProps(nextProps) {
    const { duration, autoClose, onMaskClick } = nextProps;

    if ('visible' in nextProps) {
      this.setState({
        isShow: nextProps.visible,
      });

      if (duration === 0 || this.state.timer) {
        return;
      }
      if (autoClose) {
        this.state.timer = setTimeout(() => {
          onMaskClick();
          clearTimeout(this.state.timer);
          this.setState({
            timer: null,
          });
        }, duration);
      }
    }
  }

  componentWillUnmount() {
    Events.off(this.popup, 'webkitTransitionEnd', this.animationEnd);
    Events.off(this.popup, 'webkitTransitionEnd', this.animationEnd);
  }

  animationEnd() {
    const { onClose } = this.props;

    !this.state.isShow && typeof onClose === 'function' && onClose();
  }

  render() {
    const { prefixCls, children, onMaskClick, direction, className } = this.props;
    const { isShow } = this.state;

    const cls = classnames({
      [`${prefixCls}`]: true,
    });

    const clsWrap = classnames({
      [`${prefixCls}-wrapper`]: true,
      'ui-popup-hidden': !isShow,
      [`position-${direction}`]: true,
      [className]: !!className,
    });

    return (
      <div className={cls} ref={(popup) => { this.popup = popup; }}>
        <div className={clsWrap}>
          {children}
        </div>
        {this.props.mask && <Mask visible={isShow} onClose={onMaskClick} />}
      </div>
    );
  }
}

Popup.propTypes = {
  prefixCls: PropTypes.string,
  onMaskClick: PropTypes.func,
  onClose: PropTypes.func,

};

Popup.defaultProps = {
  prefixCls: 'ui-popup',
  mask: true,
  direction: 'bottom',
  duration: 3000,
  autoClose: false,
  onMaskClick() {},
  onClose() {},
};

export default Popup;
