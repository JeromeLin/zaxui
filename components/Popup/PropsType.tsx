export default interface PropsType {
  visible?: boolean;
  shape?: 'radius';
  direction?: 'top' | 'right' | 'bottom' | 'left';
  autoClose?: boolean;
  stayTime?: number;
  animationDuration?: number;
  onClose?: () => void;
  mask?: boolean;
  maskType?: 'transparent' | 'light' | 'normal' | 'dark';
  onMaskClick?: () => void;
}
