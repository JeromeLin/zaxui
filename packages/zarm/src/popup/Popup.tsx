import * as React from 'react';
import type BasePopupProps from './PropsType';
import Portal from './Portal';

export interface PopupProps extends BasePopupProps {
  prefixCls?: string;
  className?: string;
}

const Popup = React.forwardRef<unknown, PopupProps>(
  ({ prefixCls = 'za-popup', destroy = true, visible = false, ...restProps }: PopupProps, ref) => {
    const popupRef = (ref as any) || React.createRef<typeof Portal>();

    const [renderPortal, setRenderPortal] = React.useState(false);
    const [portalVisible, setPortalVisible] = React.useState(visible);

    const handlePortalUnmount = () => {
      destroy && setPortalVisible(false);
      setRenderPortal(!destroy);
    };

    const portalRender = (
      <Portal
        ref={popupRef}
        prefixCls={prefixCls}
        visible={portalVisible}
        handlePortalUnmount={handlePortalUnmount}
        {...restProps}
      />
    );

    React.useEffect(() => {
      visible && setRenderPortal(true);
      setPortalVisible(visible);
    }, [visible, destroy]);

    return renderPortal ? portalRender : null;
  },
);

export default Popup;

// export interface PopupState {
//   renderPortal?: boolean;
//   portalVisible: boolean;
// }

// export default class Popup extends React.PureComponent<PopupProps, PopupState> {
//   static defaultProps: PopupProps = {
//     prefixCls: 'za-popup',
//     destroy: true,
//     visible: false,
//   };

//   portalRef: Portal | null;

//   static getDerivedStateFromProps(props: PopupProps, state: PopupState) {
//     if (props.visible !== state.portalVisible) {
//       if (props.visible) {
//         return {
//           renderPortal: true,
//           portalVisible: true,
//         };
//       }
//       return {
//         portalVisible: false,
//       };
//     }
//     return null;
//   }

//   constructor(props) {
//     super(props);
//     this.state = {
//       renderPortal: false,
//       portalVisible: false,
//     };
//     this.handlePortalUnmount = this.handlePortalUnmount.bind(this);
//   }

//   componentDidMount() {
//     const { visible } = this.props;
//     if (visible) {
//       this.setState({
//         renderPortal: true,
//         portalVisible: true,
//       });
//     }
//   }

//   handlePortalUnmount() {
//     const { destroy } = this.props;
//     if (destroy) {
//       this.setState({
//         renderPortal: false,
//       });
//     } else {
//       this.setState({
//         renderPortal: true,
//         portalVisible: false,
//       });
//     }
//   }

//   render() {
//     const { renderPortal, portalVisible } = this.state;
//     return (
//       renderPortal && (
//         <Portal
//           ref={(ref) => {
//             this.portalRef = ref;
//           }}
//           {...this.props}
//           visible={portalVisible}
//           handlePortalUnmount={this.handlePortalUnmount}
//         />
//       )
//     );
//   }
// }
