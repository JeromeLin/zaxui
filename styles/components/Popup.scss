$prefixCls: ui-popup;

.#{$prefixCls} {
  position: relative;

  &-wrapper {
    position: fixed;
    transform: translate(0, 0);
    transition: transform .2s;
    z-index: $zIndex-popup;
    background: #fff;

    &.position-bottom {
      left: 0;
      bottom: 0;
      width: 100%;
    }

    &.position-top {
      left: 0;
      top: 0;
      width: 100%;
    }

    &.position-left {
      left: 0;
      top: 0;
      height: 100%;
    }

    &.position-right {
      right: 0;
      top: 0;
      height: 100%;
    }
  }

  &-hidden {
    &.position-bottom {
      transform: translate(0, 100%);
    }

    &.position-top {
      transform: translate(0, -100%);
    }

    &.position-left {
      transform: translate(-100%, 0);
    }

    &.position-right {
      transform: translate(100%, 0);
    }
  }
}
