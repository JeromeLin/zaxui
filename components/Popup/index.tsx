import React, { PureComponent } from 'react';
import classnames from 'classnames';
import PropsType from './PropsType';
import Events from '../utils/events';
import Mask from '../Mask';

export interface PopupProps extends PropsType {
  prefixCls?: string;
  className?: string;
}

export default class Popup extends PureComponent<PopupProps, any> {

  static defaultProps = {
    prefixCls: 'za-popup',
    visible: false,
    mask: true,
    direction: 'bottom',
    autoClose: false,
    stayTime: 3000,
    animationDuration: 200,
    maskType: Mask.defaultProps.type,
  };

  private timer: number;
  private popup;

  constructor(props) {
    super(props);
    this.state = {
      isShow: props.visible || false,
      isMaskShow: props.visible || false,
      isPending: false,
      animationState: 'enter',
    };
  }

  componentDidMount() {
    Events.on(this.popup, 'webkitTransitionEnd', this.animationEnd);
    Events.on(this.popup, 'transitionend', this.animationEnd);
  }

  componentWillReceiveProps(nextProps) {
    clearTimeout(this.timer);

    if (nextProps.visible) {
      this.enter(nextProps);
    } else {
      this.leave();
    }
  }

  componentWillUnmount() {
    Events.off(this.popup, 'webkitTransitionEnd', this.animationEnd);
    Events.off(this.popup, 'transitionend', this.animationEnd);
  }

  enter = ({ stayTime, autoClose, onMaskClick }) => {
    this.setState({
      isShow: true,
      isMaskShow: true,
      isPending: true,
      animationState: 'enter',
    });

    if (stayTime > 0 && autoClose) {
      this.timer = setTimeout(() => {
        onMaskClick();
        clearTimeout(this.timer);
      }, stayTime);
    }
  }

  leave = () => {
    this.setState({
      isShow: false,
      isPending: true,
      animationState: 'leave',
    });
  }

  animationEnd = () => {
    const { onClose } = this.props;
    const { animationState } = this.state;

    if (animationState === 'leave') {
      this.setState({
        isMaskShow: false,
      });
      if (typeof onClose === 'function') {
        onClose();
      }
    }
  }

  renderMask = () => {
    const { mask, maskType, onMaskClick, animationDuration } = this.props;
    const { isPending, animationState, isMaskShow } = this.state;

    const maskCls = classnames({
      [`fade-${animationState}`]: isPending,
    });

    const maskStyle = {
      WebkitAnimationDuration: `${animationDuration}ms`,
      animationDuration: `${animationDuration}ms`,
    };

    return mask && (
      <Mask
        className={maskCls}
        style={maskStyle}
        visible={isMaskShow}
        type={maskType}
        onClose={onMaskClick}
      />
    );
  }

  render() {
    const { prefixCls, className, animationDuration, direction, children } = this.props;
    const { isShow } = this.state;

    const popupCls = classnames(`${prefixCls}`, className, {
      [`${prefixCls}-hidden`]: !isShow,
    });
    const wrapCls = classnames(`${prefixCls}-wrapper`, `${prefixCls}-wrapper-${direction}`);

    const wrapStyle = {
      WebkitTransitionDuration: `${animationDuration}ms`,
      transitionDuration: `${animationDuration}ms`,
    };

    return (
      <div className={popupCls} ref={(popup) => { this.popup = popup; }}>
        <div className={wrapCls} style={wrapStyle}>
          {children}
        </div>
        {this.renderMask()}
      </div>
    );
  }
}
