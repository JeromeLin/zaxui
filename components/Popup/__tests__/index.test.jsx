import React from 'react';
import { render } from 'enzyme';
import toJson from 'enzyme-to-json';
import Popup from '../index';

describe('Popup', () => {
  it('renders correctly', () => {
    const onMaskClick = jest.fn();
    const onClose = jest.fn();
    const wrapper = render(
      <Popup
        visible
        direction="bottom"
        onMaskClick={onMaskClick}
        onClose={onClose}>
        foo
      </Popup>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
