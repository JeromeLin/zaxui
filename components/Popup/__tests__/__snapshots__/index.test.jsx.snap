// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup renders correctly 1`] = `
<div
  class="za-popup"
>
  <div
    class="za-popup-wrapper za-popup-wrapper-bottom"
    style="-webkit-transition-duration:200ms;transition-duration:200ms;"
  >
    foo
  </div>
  <div
    class="za-mask normal"
  />
</div>
`;
